### PR TITLE
[Feature] hunyuanimage support flash attn

### PR DESCRIPTION
**Pull request by Bounty-hunter**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Re-implement flash attention mechanism for https://github.com/vllm-project/vllm-omni/pull/1975

### changes
(1) **Introduce segmented attention utility**
Add diffusion/attention/backends/utils/piecewise_attn.py to enable attention execution in a segment-wise manner.
Example:
```
Input sequence:
[text_prompt1, image_prompt1, text_prompt2, image_prompt2]

Execution order:
attn(text_prompt1) → attn(image_prompt1) → attn(text_prompt2) → attn(image_prompt2)
```
(2) Extend FlashAttention backend
Modify the FlashAttention to support segment-based execution

(3) Integration for Hunyuan Image
For Hunyuan Image, construct and pass the required segmentation parameters (e.g., image spans) to drive the segmented attention workflow.

## Test Plan
(1) unitest for segment execute flow 
(1) compatible with tp, sp，cfg parallel
(2) performance improvement

## Test Result
**pytest**
image

**H800**
```
python ../vllm-omni/examples/offline_inference/text_to_image/text_to_image.py --stage-configs-path ../vllm-omni/vllm_omni/model_executor/stage_configs/hunyuan_image3_t2i.yaml --model /scratch/fq9hpsac/d00806799/model/hunyuanimage-3.0 --prompt "A brown and white dog is running on the grass" --output output_${TASK}.png --num-inference-steps 50 --guidance-scale 5.0 --seed 1234 --enable-diffusion-pipeline-profiler --log-stats > ${TASK}.out 2>&1 &
```

| case | backend | time | speed up| layers[0].forward | layers[0].self_attn | layers[0].mlp | output |
|-----|-----|-----|-----|-----|-----|-----|-----|
| TP4 | sdpa |14.6s | base | 0.45s | **0.17s** | 0.21s |  output_tp4_sdpa  |
| TP4 | fa | 13.4s | **9%**  | 0.42s | **0.14s** | 0.21s |  output_tp4_fa  |
| TP2+ fp8 + sp2 | sdpa |11.8s | base | 0.365s | **0.18s** | 0.14s |  output_tp2_fp8_sp2_sdpa  |
| TP2+ fp8 + sp2 | fa | 10.5s | **12.3%**  | 0.33s | **0.15s** | 0.14s |  output_tp2_fp8_sp2_fa  |
| TP2+ fp8 + cfgP2 | sdpa | 9.28 s | base | 0.30s | **0.11s** | 0.14s |  output_tp2_fp8_cfg2_sdpa  |
| TP2+ fp8 + cfgP2 | fa | 8.11s | **14.4%**  | 0.26s | **0.08s** | 0.14s |  output_tp2_fp8_cfg2_fa  |

**L20x**
| case | backend | DIT time | speed up | output |
|-----|-----|-----|-----|-----|
| it2i | sdpa |23.4s | base | | 
| it2i | flash_attn |27.24s | 16% | | 
| DIT-tp4| sdpa | 24.96 |  | tp4 | 
| DIT-tp4| flash_attn | 22.71 | 9.9% | tp4 |
| DIT-tp2sp2| sdpa | 27.78 | base | tp2sp2 | 
| DIT-tp2sp2| flash_attn | 25.44 | 9.1% | tp2sp2 | 
| DIT-tp2cfgp2| spda| 24.80 | base | tp2cfgp2 | 
| DIT-tp2cfgp2| flash_attn | 22.88 | 8.7% | tp2cfgp2 | 

**The absolute latency reduction from Flash Attention remains roughly constant, but as the overall end-to-end latency decreases, its relative speedup becomes larger.** 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by lishunyang12** [comment]
Which flash attn you are wiring to this mode? Flash attention 3?

**Review by hsliuustc0106 — COMMENTED** [review]
Ready for full review when draft status removed.

Note: Pre-commit check is currently failing. Please run `pre-commit run --all-files` and fix any linting issues before removing draft status.

Also, the PR description has detailed design documentation but the Test Plan and Test Result sections are empty. Please populate these with evidence of testing (unit tests for segmented_prefix_attn, e2e parity tests, and performance benchmarks) as outlined in your design doc.

**Comment by Bounty-hunter** [comment]
> Which flash attn you are wiring to this mode? Flash attention 3?
It is version-agnostic and supports both FlashAttention v2 and v3.

**Comment by Bounty-hunter** [comment]
> Ready for full review when draft status removed.
> 
> Note: Pre-commit check is currently failing. Please run `pre-commit run --all-files` and fix any linting issues before removing draft status.
> 
> Also, the PR description has detailed design documentation but the Test Plan and Test Result sections are empty. Please populate these with evidence of testing (unit tests for segmented_prefix_attn, e2e parity tests, and performance benchmarks) as outlined in your design doc.

Draft status removed, please help review.

File: vllm_omni/diffusion/attention/backends/utils/segmented.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    value: torch.Tensor,
+    image_spans: list[list[tuple[int, int]]],  # outer length B
+    q_global_positions: torch.Tensor,  # (B, Sq)
+    softmax_scale: float,
+    attn_func,
+) -> torch.Tensor:
+    B, Sq, H, D = query.shape
+    if not _is_homogeneous_batch(image_spans, q_global_positions):
+        raise NotImplementedError(
+            "segmented_attn requires a homogeneous batch (identical image_spans and q_global_positions across samples)."
```
[the remark this answers is not in the document]
Comment (Bounty-hunter) [reply vllm_omni/diffusion/attention/backends/utils/segmented.py]: hunyuan currently only support batch with (positive, negtive prompt), witch with the same image_spans slices.

File: vllm_omni/diffusion/attention/backends/utils/segmented.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    for a, e in spans:
+        if cur >= e:
+            continue
+        elif cur < a:
+            segs.append(Segment(cur, min(a, q_end), "text"))
+            cur = a
+
+        if cur < q_end:
+            segs.append(Segment(a, e, "image"))
```
[the remark this answers is not in the document]
Comment (Bounty-hunter) [reply vllm_omni/diffusion/attention/backends/utils/segmented.py]: There is no case where q_start in (a, e)

**Comment by Bounty-hunter** [comment]
@nussejzz PTAL

File: vllm_omni/model_executor/stage_configs/hunyuan_image3_t2i.yaml [hunk 3]
```diff
@@ -16,7 +16,7 @@ stage_args:
       distributed_executor_backend: "mp"
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_configs/hunyuan_image3_t2i.yaml:16]: @Bounty-hunter PTAL, this yaml should be removed later

Comment (Bounty-hunter) [reply vllm_omni/model_executor/stage_configs/hunyuan_image3_t2i.yaml:16]: Got it. This PR does not depend on this YAML change. Since it will be removed later, I’ll drop this change to avoid confusion.

File: vllm_omni/diffusion/attention/backends/utils/segmented.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            cur = e
```
Comment (SamitHuang) [inline vllm_omni/diffusion/attention/backends/utils/segmented.py]: This append uses `a` and `e` without clamping to `[q_start, q_end]`. If `a < q_start`, this causes `s - query_offset` in `segmented_attn` to become negative, leading to PyTorch out-of-bounds slice semantics that silently fail (e.g., producing empty tensors and leaving the output as zeros). Change this to `Segment(max(a, cur), min(e, q_end), "image")` and update `cur = min(e, q_end)`.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 5]
```diff
@@ -668,10 +662,16 @@ def _prepare_attention_mask_for_generation(
...
         ]
         attention_mask = torch.ones(seq_len, seq_len, dtype=torch.bool).tril(diagonal=0).repeat(bsz, 1, 1)
+        image_spans: list[list[tuple[int, int]]] = [[] for _ in range(bsz)]
         for i in range(bsz):
             for j, image_slice in enumerate(batch_image_slices[i]):
                 attention_mask[i, image_slice, image_slice] = True
+                start = image_slice.start if image_slice.start is not None else 0
+                stop = image_slice.stop if image_slice.stop is not None else seq_len
+                assert start < stop, f"Invalid image slice: {image_slice}"
+                image_spans[i].append((int(start), int(stop)))
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py]: Ensure that `image_spans[i]` is sorted by start index before appending to `model_kwargs['image_spans']`, as `build_segments` assumes a monotonic progression of spans. If `batch_image_slices` is already guaranteed to be sorted, consider adding a comment or assertion here.

File: tests/diffusion/attention/test_piecewise_attn.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    v_ = value.transpose(1, 2)
+    out = F.scaled_dot_product_attention(q_, k_, v_, attn_mask=mask_q, scale=softmax_scale)
+    return out.transpose(1, 2).contiguous()
+
+
+SPAN_CASES = [
+    pytest.param([], id="no-spans"),
+    pytest.param([(0, 10)], id="span-at-start"),
+    pytest.param([(10, 30), (54, 64)], id="multi-spans"),
+]
```
Comment (SamitHuang) [inline tests/diffusion/attention/test_piecewise_attn.py:61]: Add a test case where the `q_start` falls directly inside an image span (e.g., `q_start=15, q_end=30, span=[10, 20]`). This would catch the out-of-bounds slice bug in `build_segments` where `a < q_start` causes a negative start index during query tensor slicing.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
This PR introduces segmented Flash Attention for HunyuanImage, which is a great approach to handling mixed causal/full masks efficiently. However, there is a critical out-of-bounds slice bug in how query segments are computed. When `a < q_start` or `q_start` falls directly inside an image span, the resulting segment indices produce negative start indices for `query` slicing. In PyTorch, these map to indices counting from the end of the sequence, leading to silent empty slices or corrupted attention outputs. I've left a few inline comments to address this.

File: vllm_omni/diffusion/attention/backends/utils/segmented.py (hunk 4 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/attention/backends/utils/segmented.py]: An image span must either be entirely within interval [q_start, q_end] or entirely outside interval [q_start, q_end]; We add assert to verify it.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py (hunk 5 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py]: we add Sort after image_spans[i] initialization

File: tests/diffusion/attention/test_piecewise_attn.py (hunk 6 above)
Comment (Bounty-hunter) [reply tests/diffusion/attention/test_piecewise_attn.py:61]: added, and verify the assertion.

**Review by SamitHuang — APPROVED** [review]
LGTM

**Review by princepride — APPROVED** [review]
LGTM

File: vllm_omni/diffusion/attention/backends/utils/segmented.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""
+Segmented attention for mixed causal / bidirectional masks
+(e.g. HunyuanImage3: causal outside image spans, bidirectional inside).
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/utils/segmented.py]: Does it only apply to hunyuan image 3?

**Comment by lishunyang12** [comment]
Will it supersede https://github.com/vllm-project/vllm-omni/pull/1975?

**Review by hsliuustc0106 — APPROVED** [review]
## Blocker scan

| Category | Result |
|---|---|
| Correctness | PASS |
| Reliability/Safety | PASS |
| Breaking Changes | PASS |
| Test Coverage | PASS |
| Documentation | PASS |
| Security | PASS |

## SamitHuang's out-of-bounds bug — verified fixed

Traced through `build_segments` for the three critical cases:

- **q_start inside span** (`q=15..30, span=[10,20]`): the guard `assert a == cur` fires → `AssertionError`. Backed by `test_segmented_qstart_inside_span_asserts`.
- **Span fully before q_start** (`q=15..30, span=[5,10]`): `cur >= e` skips the span → pure text segment. Backed by `test_segmented_span_fully_before_qstart`.
- **Normal case** (`q=0..64, spans=[(10,30),(54,64)]`): produces `text[0,10] → image[10,30] → text[30,54] → image[54,64]` with correct causal/full flags. Backed by parametrized `test_segmented_matches_full`.

The K/V prefix `[:e]` for text segments after image spans correctly includes the preceding image tokens (e.g., `text[30,54]` uses `K[:,:54]` which covers `image[10,30]`). Causal mask within the text segment then enforces the correct per-position visibility.

## Other notes

- `image_spans[i].sort(key=lambda x: x[0])` in the pipeline ensures monotonic span ordering for `build_segments`.
- `_is_homogeneous_batch` correctly raises `NotImplementedError` for heterogeneous batches — fine for HunyuanImage3's CFG batch pattern.
- The `s.start : s.stop` indexing in `_split_model_kwargs_for_cfg_parallel` correctly handles `image_spans` as a list-of-lists (distinct from the tensor indexing path).
- Removing the hardcoded `DIFFUSION_ATTENTION_BACKEND=TORCH_SDPA` env var is the right call — it was blocking FlashAttention usage.

LGTM.

**Comment by Bounty-hunter** [comment]
> Will it supersede #1975?

yes

File: vllm_omni/diffusion/attention/backends/utils/segmented.py (hunk 7 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/attention/backends/utils/segmented.py]: No. -- Any model can using the segment attention by invoke `Attention` with the corresponding metadata (e.g., image spans).

**Review by gcanlin — APPROVED** [review]
